core/src/main/java/io/grpc/ClientInterceptors.java

```diff
@@ -204,4 +218,195 @@ public final void start(Listener<RespT> responseListener, Metadata headers) {
+    }
+
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+            MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+      return new RetryingCall<ReqT, RespT>(method, callOptions, next, retryPolicy);
+    }
+
+    private class RetryingCall<ReqT, RespT> extends ClientCall<ReqT, RespT> {
+      private final MethodDescriptor<ReqT, RespT> method;
+      private final CallOptions callOptions;
+      private final Channel channel;
+      private final ScheduledExecutorService scheduledExecutor;
+      private Listener<RespT> responseListener;
+      private Metadata requestHeaders;
+      private ReqT requestMessage;
+      private boolean compressionEnabled;
+      private final Queue<AttemptListener> attemptListeners =
+              new ConcurrentLinkedQueue<AttemptListener>();
+      private volatile AttemptListener latestResponse;
+      private final RetryPolicy retryPolicy;
+
+      RetryingCall(MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel channel,
+                   RetryPolicy retryPolicy) {
+        this.method = method;
+        this.callOptions = callOptions;
+        this.channel = channel;
+        this.retryPolicy = retryPolicy;
+        this.scheduledExecutor = SharedResourceHolder.get(TIMER_SERVICE);
+      }
+
+      @Override
+      public void start(Listener<RespT> listener, Metadata headers) {
+        checkState(attemptListeners.isEmpty());
+        checkState(responseListener == null);
+        checkState(requestHeaders == null);
+        responseListener = listener;
+        requestHeaders = headers;
+        ClientCall<ReqT, RespT> firstCall = channel.newCall(method, callOptions);
+        AttemptListener attemptListener = new AttemptListener(firstCall);
+        attemptListeners.add(attemptListener);
+        firstCall.start(attemptListener, headers);
+      }
+
+      @Override
+      public void request(int numMessages) {
+        lastCall().request(numMessages);
+      }
+
+      @Override
+      public void cancel() {
+        for (AttemptListener attempt : attemptListeners) {
+          attempt.call.cancel();
+        }
+      }
+
+      @Override
+      public void halfClose() {
+        lastCall().halfClose();
+      }
+
+      @Override
+      public void sendMessage(ReqT message) {
+        checkState(requestMessage == null);
+        requestMessage = message;
+        lastCall().sendMessage(message);
+      }
+
+      @Override
+      public boolean isReady() {
+        return lastCall().isReady();
+      }
+
+      @Override
+      public void setMessageCompression(boolean enabled) {
+        compressionEnabled = enabled;
+        lastCall().setMessageCompression(enabled);
+      }
+
+      private void maybeRetry(AttemptListener attempt) {
+        Status status = attempt.responseStatus;
+        if (!retryPolicy.isRetryable(status, method, callOptions)) {
+          return;
```

Comment:
Weird that it "worked" in my quick testing, but yeah definitely it should finish call. I think that name can stay? Fixed.